politeiawww/codetracker/github/update.go

```diff
@@ -0,0 +1,171 @@
+package github
+
+import (
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	cms "github.com/decred/politeia/politeiawww/api/cms/v1"
+	"github.com/decred/politeia/politeiawww/codetracker/github/api"
+	"github.com/decred/politeia/politeiawww/codetracker/github/database"
+	"github.com/decred/politeia/politeiawww/codetracker/github/database/cockroachdb"
+)
+
+// github implements the Tracker interface for github.
+type github struct {
+	tc     *api.Client
+	codedb database.Database
+}
+
+// Setup creates a new github tracker that saves is able to communicate with the
+// Github user/PR/issue API.
+func New(apiToken, host, rootCert, cert, key string) (*github, error) {
+	var err error
+	g := &github{}
+	g.tc = api.NewClient(apiToken)
+	g.codedb, err = cockroachdb.New(host, rootCert, cert, key)
+	if err == database.ErrNoVersionRecord || err == database.ErrWrongVersion {
+		log.Errorf("New DB failed no version, wrong version: %v\n", err)
+		return nil, err
+	} else if err != nil {
+		log.Errorf("New DB failed: %v\n", err)
+		return nil, err
+	}
+	err = g.codedb.Setup()
+	if err != nil {
+		log.Errorf("codeDB setup failed: %v", err)
+		return nil, err
+	}
+	return g, nil
+}
+
+// Update fetches all repos from the given organization and updates all
+// users' information once the info is fully received.  If repoRequest is
+// included then only that repo will be fetched and updated, typically
+// used for speeding up testing.
+func (g *github) Update(org string, repoRequest string) error {
+	// Fetch the organization's repositories
+	repos, err := g.tc.FetchOrgRepos(org)
+	if err != nil {
+		err = fmt.Errorf("FetchOrgRepos: %v", err)
+		return err
+	}
+
+	for _, repo := range repos {
+		// Allow for a repo argument for testing expediting.
+		if repoRequest != "" && repo.Name != repoRequest {
+			continue
+		}
+
+		log.Infof("%s", repo.Name)
+		log.Infof("Syncing %s", repo.FullName)
+
+		// Grab latest sync time
+		prs, err := g.tc.FetchPullsRequest(org, repo.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, pr := range prs {
+			var prNum [8]byte
+			binary.LittleEndian.PutUint64(prNum[:], uint64(pr.Number))
+
+			apiPR, err := g.tc.FetchPullRequest(org, repo.Name, pr.Number)
+			if err != nil {
+				return err
+			}
+			dbPullRequest, err := convertAPIPullRequestToDbPullRequest(apiPR, *repo, org)
+			if err != nil {
+				log.Errorf("error converting api PR to database: %v", err)
+				continue
+			}
+			dbPR, err := g.codedb.PullRequestByURL(dbPullRequest.URL)
+			if err != nil {
+				if err == database.ErrNoPullRequestFound {
```

Comment:
This extra nesting isn't necessary.  The more nested code is the harder it is to read and reason about.  Just do:

```
if err == database.ErrNoPullRequestFound {
  ...
} else if err != nil {
  ...
}
```

This code is a little rough.  Lots of going over the column limit.  Lots of nesting inside of nesting.  Can you go through these files and clean it up a bit.  Maybe break some of it off into its own function to improve readability.  For example, you got nested for loops here, but if you broke the PR handling code into its own `g.handlePR()` function it would go a long way to improving the readability.  Stuff like that.